pkg/neonvm/controllers/vm_controller.go

```diff
@@ -802,6 +837,217 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
+
+	return certSecret, nil
+}
+
+func (r *VMReconciler) doReconcileCertificateSecret(ctx context.Context, vm *vmv1.VirtualMachine, certSecret *corev1.Secret) (*corev1.Secret, error) {
+	log := log.FromContext(ctx)
+
+	// Check if the TLS private key temporary secret exists, if not create a new one
+	tmpKeySecret := &corev1.Secret{}
+	err := r.Get(ctx, types.NamespacedName{Name: fmt.Sprintf("%s-tmp", vm.Status.TLSSecretName), Namespace: vm.Namespace}, tmpKeySecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "Failed to get vm-runner Secret")
+		return nil, err
+	} else if err != nil /* not found */ {
+		tmpKeySecret, err = r.createTlsTmpSecret(ctx, vm)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	key, err := pki.DecodePrivateKeyBytes(tmpKeySecret.Data[corev1.TLSPrivateKeyKey])
+	if err != nil {
+		log.Error(err, "Failed to decode private key")
+	}
+
+	// Check if the TLS certificate already exists, if not create a new one
+	certificateReq := &certv1.CertificateRequest{}
+	err = r.Get(ctx, types.NamespacedName{Name: vm.Name, Namespace: vm.Namespace}, certificateReq)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "Failed to get vm-runner CertificateRequest")
+		return nil, err
+	} else if err != nil /* not found */ {
+		certificateReq, err = r.createCertificateRequest(ctx, vm, key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if len(certificateReq.Status.Certificate) == 0 {
+		// we cannot yet update the cert secret.
+		// return it untouched.
+		return certSecret, nil
+	}
+
+	// we have a certificate and the corresponding private key
+	// create/update the proper certificate secret and delete the tmp secret
+	if certSecret == nil {
+		certSecret, err = r.createTlsSecret(ctx, vm, key, certificateReq)
+		if err != nil {
+			return nil, err
+		}
+	} else if !reflect.DeepEqual(certificateReq.Status.Certificate, certSecret.Data[corev1.TLSCertKey]) {
+		if err := r.updateTlsSecret(ctx, vm, key, certificateReq, certSecret); err != nil {
+			return nil, err
+		}
+	}
+
+	// parse the certificate for the expiration time.
+	certs, err := pki.DecodeX509CertificateChainBytes(certSecret.Data[corev1.TLSCertKey])
+	if err != nil {
+		log.Error(err, "Failed to parse VM certificate")
+		return nil, err
+	}
+	vm.Status.TLSRenewAt.Time = certs[0].NotAfter.Add(-vm.Spec.TLS.RenewBefore.Duration)
+
+	r.deleteTmpSecret(ctx, vm, tmpKeySecret, certificateReq)
+
+	return certSecret, nil
+}
+
+func (r *VMReconciler) createTlsTmpSecret(ctx context.Context, vm *vmv1.VirtualMachine) (*corev1.Secret, error) {
+	log := log.FromContext(ctx)
+
+	// create a new key for this VM
+	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		log.Error(err, "Failed to generate private key for VirtualMachine")
+		return nil, err
+	}
+
+	// Define the secret
+	tmpKeySecret, err := r.tmpKeySecretForVirtualMachine(vm, key)
+	if err != nil {
+		log.Error(err, "Failed to define new Secret resource for VirtualMachine")
```

Comment:
suggestion: Similar to above, we have multiple places where we create secrets for VMs (both here and for SSH keys) - let's have some more context in this log line?